backend.native/tests/runtime/text/utf8.kt

```diff
@@ -353,13 +434,53 @@ fun test8to16CustomBorders() {
     checkUtf8to16Throws("\uFFFD1", intArrayOf(-16, -97, -104, '1'.toInt(), '2'.toInt()), 0, 4)
     checkUtf8to16Throws("\uFFFD2", intArrayOf('1'.toInt(), -16, -97, -104, '2'.toInt(), '3'.toInt()), 1, 4)
     checkUtf8to16Throws("2\uFFFD", intArrayOf('1'.toInt(), '2'.toInt(), -16, -97, -104),  1, 4)
+}
+// endregion
 
-    // Index out of bound
-    checkThrows(IndexOutOfBoundsException::class, "Hello") { helloArray.stringFromUtf8OrThrow(-1, 4)}
-    checkThrows(IndexOutOfBoundsException::class, "Hello") { helloArray.stringFromUtf8OrThrow(5, 10)}
-    checkThrows(IndexOutOfBoundsException::class, "Hello") { helloArray.stringFromUtf8OrThrow(2, 10)}
-    checkThrows(IndexOutOfBoundsException::class, "Hello") { helloArray.stringFromUtf8OrThrow(10, 0)}
-    checkThrows(IllegalArgumentException::class, "Hello") { helloArray.stringFromUtf8OrThrow(3, -2)}
+// region zero-terminated UTF-8 -> UTF-16 conversion
+fun testZeroTerminated8To16() {
+    // Valid strings.
+    checkValidZeroTerminatedUtf8to16("Hell", intArrayOf('H'.toInt(), 'e'.toInt(), 'l'.toInt(), 'l'.toInt(), 0, 'o'.toInt()))
+    checkValidZeroTerminatedUtf8to16("При", intArrayOf(-48, -97, -47, -128, -48, -72, 0, -48, -78, 0, -48, -75, -47, -126))
+    checkValidZeroTerminatedUtf8to16("\uD800\uDC00", intArrayOf(-16, -112, -128, -128, 0, -16, -112, -128, -128))
+    checkValidZeroTerminatedUtf8to16("", intArrayOf(0, 'H'.toInt()))
+
+    // Test manual conversion with replacement.
+    // Incorrect UTF-8 lead character.
+    checkZeroTerminatedUtf8to16Replacing("\uFFFD", intArrayOf(-1, 0, '1'.toInt()))
+
+    // Incomplete codepoint.
+    checkZeroTerminatedUtf8to16Replacing("\uFFFD", intArrayOf(-16, -97, -104, 0, '1'.toInt()))
+    checkZeroTerminatedUtf8to16Replacing("\uFFFD1", intArrayOf(-16, -97, -104, '1'.toInt(), 0, -16, -97, -104))
+
+    // Test manual conversion with exception throwing
+    // Incorrect UTF-8 lead character -> throw.
+    checkZeroTerminatedUtf8to16Throws("\uFFFD", intArrayOf(-1, 0, '1'.toInt()))
+
+    // Incomplete codepoint -> throw.
+    checkZeroTerminatedUtf8to16Throws("\uFFFD", intArrayOf(-16, -97, -104, 0, '1'.toInt()))
+    checkZeroTerminatedUtf8to16Throws("\uFFFD1", intArrayOf(-16, -97, -104, '1'.toInt(), 0, -16, -97, -104))
+}
+
+fun testZeroTerminated8To16CustomBorders() {
+    val array = intArrayOf('a'.toInt(), 'a'.toInt(), 'a'.toInt(), 0, 'b'.toInt(), 'b'.toInt(), 'b'.toInt(), 0)
+    checkValidZeroTerminatedUtf8to16("aaa", array, 0, 5)
+    checkValidZeroTerminatedUtf8to16("a", array, 2, 2)
+    checkValidZeroTerminatedUtf8to16("", array, 3, 2)
+    checkValidZeroTerminatedUtf8to16("bb", array, 4, 2)
```

Comment:
Here the terminating zero is searched starting from the given `start` index. Do we want to persist this behaviour in the new API?

Reply:
Yes, it seems logical: when we're working with a subrange of an array, we do not look to the contents out of this subrange.